be/src/service/service_be/lake_service.cpp

```diff
@@ -4,27 +4,146 @@
+    }
+
+    if (log.has_op_schema_change()) {
+        return Status::NotSupported("does not support apply schema change log yet");
+    }
+
+    return Status::OK();
+}
+
 void LakeServiceImpl::publish_version(::google::protobuf::RpcController* controller,
                                       const ::starrocks::lake::PublishVersionRequest* request,
                                       ::starrocks::lake::PublishVersionResponse* response,
                                       ::google::protobuf::Closure* done) {
     brpc::ClosureGuard guard(done);
     (void)controller;
-    (void)request;
-    response->mutable_status()->set_status_code(TStatusCode::NOT_IMPLEMENTED_ERROR);
+    if (!request->has_version()) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("missing version");
+        return;
+    }
+    if (request->txn_ids_size() == 0) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("missing txn_ids");
+        return;
+    }
+    if (request->tablet_ids_size() == 0) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("missing tablet_ids");
+        return;
+    }
+    int64_t base_version = request->version() - request->txn_ids_size();
+    if (base_version < 1) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("invalid version");
+        return;
+    }
+    // Will not update status code since here, only failed_tablets will be updated.
+    response->mutable_status()->set_status_code(TStatusCode::OK);
+
+    // TODO move the execution to TaskWorkerPool
+    for (const auto& tablet_id : request->tablet_ids()) {
+        auto tablet = _env->lake_tablet_manager()->get_tablet(tablet_id);
+        if (!tablet.ok()) {
+            LOG(WARNING) << "Fail to get tablet " << tablet_id << ": " << tablet.status();
+            response->add_failed_tablets(tablet_id);
+            continue;
+        }
+        auto base_metadata = tablet->get_metadata(base_version);
+        if (base_metadata.status().is_not_found() && tablet->get_metadata(request->version()).ok()) {
+            // the current tablet has been successfully published before, skip to the next tablet
+            continue;
+        } else if (!base_metadata.ok()) {
+            LOG(WARNING) << "Fail to get " << tablet->metadata_path(base_version) << ": " << base_metadata.status();
+            response->add_failed_tablets(tablet_id);
+            continue;
+        }
+
+        // make a copy of metadata
+        auto new_metadata = std::make_shared<lake::TabletMetadata>(**base_metadata);
+        new_metadata->set_version(request->version());
+
+        bool succeeded = false;
+        for (auto txn_id : request->txn_ids()) {
```

Comment:
Does the tablet metadata between base version and publish version need to be generated?

Reply:
No

Reply:
There is no need, at least now,  to generate multiple versions in a single publish version request.

Reply:
Do not consider time travel?